README.md
# AngularSample

This project was generated with [Angular CLI](https://github.com/angular/angular-cli) version 11.2.9.

## Development server

Run `ng serve` for a dev server. Navigate to `http://localhost:4200/`. The app will automatically reload if you change any of the source files.

## Code scaffolding

Run `ng generate component component-name` to generate a new component. You can also use `ng generate directive|pipe|service|class|guard|interface|enum|module`.

## Build

Run `ng build` to build the project. The build artifacts will be stored in the `dist/` directory. Use the `--configuration production` flag for a production build.

## Running unit tests

Run `ng test` to execute the unit tests via [Karma](https://karma-runner.github.io).

## Running end-to-end tests

Run `ng e2e` to execute the end-to-end tests via [Protractor](http://www.protractortest.org/).

## Further help

To get more help on the Angular CLI use `ng help` or go check out the [Angular CLI Overview and Command Reference](https://angular.io/cli) page.

## Build data

Run `tsc src/data/createData`
Run `node src/data/createData`


ChangeLog.md

# ChangeLog

2022/04/27

1. Router: Add Scroll to top
2. Router: width style move into separate component
3. Router: add height changed animation

2022/03/06

1. Implement Kelly criterion page

2022/03/04

1. Add Kelly criterion page
2. Add MathjaxModule
3. Set outlet container width
4. Change fade parameter

2022/03/03

1. Add path 'hide' can hide toolbar and sidenav

2021/09/22

1. Change stepper style
2. Modify investment result component

2021/09/14

1. Change data
2. Create investment-result-and-source
3. Refine investment page

2021/09/13

1. Add route fade animation
2. Change global-variables and add global-methods
3. Change LINK_INFOS
4. remove toolbar component

2021/09/09

1. sidenav and toolbar add responsive design for mobile
2. Add title and font Lato
3. Change logic of 'leftSidenav.opened'

2021/09/08

1. Add maximumError size

2021/09/06

1. Merge from other project
2. Remove Github icon color then it can be changed color by style
3. Add some project link
4. Add azure routing: web.config
5. Move directory

2021/09/05

1. Implement versus page (part3)
2. Implement result component (part1)
3. leftSidenav openedChange trigger window.resize event

2021/09/02

1. Implement versus page (part2)
2. Add sidenav link active color

2021/09/01

1. Move methods to services
2. Implement versus page (part)
3. Responsive sidenav mode

2021/08/29

1. Add sidenave and versus page(empty)

2021/08/27

1. Change routes

2021/08/23

1. Implement step 2 data range selection

2021/08/22

1. Create result page and change class name

2021/08/19

1. finish investment stepper functions

2021/08/18

1. Add stepper
2. Update angular material version

2021/08/17

1. Change data format and add data
2. Load and show all data
3. Update angular version

2021/08/16

1. Add Dollar-Cost Averaging and Value Averaging

2021/08/15

1. Create branch to implement investment page
2. Add investment page and some test code


src/app/app.component.ts
import { ChangeDetectorRef, Component, OnDestroy, ViewChild } from '@angular/core';
import { ActivatedRoute, ChildrenOutletContexts, NavigationEnd, Router } from '@angular/router';
import { MediaMatcher } from '@angular/cdk/layout';
import { DomSanitizer } from '@angular/platform-browser';
import { filter, map } from 'rxjs/operators';

import { MatSidenav } from '@angular/material/sidenav';
import { MatIconRegistry } from '@angular/material/icon';

import { GlobalMethodsService } from './services/global-methods.service';
import { LinkInfo } from './classes/LinkInfo';
import { FADE_ANIMATION, LINK_INFOS, MENU_BTN_ANIMATION } from './global-variables';


@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.scss'],
  animations: [FADE_ANIMATION, MENU_BTN_ANIMATION]
})
export class AppComponent implements OnDestroy {
  title = 'angular-sample';

  @ViewChild('leftSidenav') leftSidenav!: MatSidenav;
  @ViewChild('container') navContainer!: any;

  private _mobileQueryListener: (e: MediaQueryListEvent) => void;
  routerHasNav: boolean = false;

  get isMobile(): boolean {
    return this.gvs.isMobile;
  }
  get linkInfos(): LinkInfo[] {
    return LINK_INFOS;
  }
  get isShownMenu(): boolean {
    return this.isMobile || this.routerHasNav;
  }

  hideLayout: boolean = false;

  constructor(
    route: ActivatedRoute,
    private router: Router,
    private contexts: ChildrenOutletContexts,
    changeDetectorRef: ChangeDetectorRef,
    media: MediaMatcher,
    matIconRegistry: MatIconRegistry,
    domSanitizer: DomSanitizer,
    private gvs: GlobalMethodsService) {

    this.initializeEvents(router, route);

    // 20210909: detect window width
    gvs.mobileQuery = media.matchMedia('(max-width: 600px)');
    this._mobileQueryListener = (e: MediaQueryListEvent) => changeDetectorRef.detectChanges();
    gvs.mobileQuery.addEventListener('change', this._mobileQueryListener);

    matIconRegistry.addSvgIcon(
      "githubicon",
      domSanitizer.bypassSecurityTrustResourceUrl("../assets/GitHub.svg")
    );
  }

  ngOnDestroy(): void {
    this.gvs.mobileQuery.removeEventListener('change', this._mobileQueryListener);
  }

  ngAfterViewInit(): void {
    // 2021/09/06: leftSidenav openedChange trigger resize to resize chart

    //console.log('ngAfterViewInit');
    this.leftSidenav.openedChange.subscribe(e => {
      //console.log('subscribe');
      window.dispatchEvent(new Event('resize'));
    });
  }

  initializeEvents(router: Router, route: ActivatedRoute) {

    let navEnd$ = router.events.pipe(
      filter(e => e instanceof NavigationEnd),
      map(e => e as NavigationEnd),
    );
    navEnd$.subscribe(event => {
      //console.log("--NavigationEnd");
      this.scrollToTop();

      this.updateSidenavVisibilty(event);

      // check if url has '/hide'
      route.firstChild?.paramMap.subscribe({
        next: (params) => {
          this.hideLayout = params.get('hideLayout') === 'hide';
        }
      });
    });
  }

  getRouteAnimationData() {
    return this.contexts.getContext('primary')?.route?.snapshot?.data?.['animation'];
  }

  updateSidenavVisibilty(event: NavigationEnd) {
    // console.log(event);

    this.routerHasNav = this.gvs.currentRouterHasNav(this.router);
    this.leftSidenav.opened = !this.isMobile && this.routerHasNav;
  }

  scrollToTop() {
    this.navContainer.scrollTo({
      top: 0,
      left: 0,
      behavior: 'smooth'
    });
  }
}
